dataverse-webapp/src/main/java/edu/harvard/iq/dataverse/engine/command/impl/ReturnDatasetToAuthorCommand.java

```diff
@@ -43,13 +43,13 @@ public Dataset execute(CommandContext ctxt)  {
         dataset.getEditVersion().setLastUpdateTime(getTimestamp());
         dataset.setModificationTime(getTimestamp());
 
-        ctxt.engine().submit(new RemoveLockCommand(getRequest(), getDataset(), DatasetLock.Reason.InReview));
         WorkflowComment workflowComment = new WorkflowComment(dataset.getEditVersion(), WorkflowComment.Type.RETURN_TO_AUTHOR, comment, (AuthenticatedUser) this.getUser());
         ctxt.datasets().addWorkflowComment(workflowComment);
 
         updateDatasetUser(ctxt);
         ctxt.em().flush();
         Dataset savedDataset = ctxt.em().merge(dataset);
+        ctxt.engine().submit(new RemoveLockCommand(getRequest(), getDataset(), DatasetLock.Reason.InReview));
```

Comment:
Definitely it is a step in the right direction. What I would also try to do is apply a flush after the merge of dataset instead of before.
```
        ctxt.em().flush();
        Dataset savedDataset = ctxt.em().merge(dataset);
```
->
```
        ctxt.em().flush();
        Dataset savedDataset = ctxt.em().merge(dataset);
```
That way we will be more sure that everything is fine before execution of `RemoveLockCommand`

There still can be a chance of making `inconsistent` state of application when sendNotification or indexDataset will throw an exception (the whole transaction will be rolled back, but lock will turned off). But I think it is debatable whether we should worry about that.